File: JavaLibrary/src/main/java/com/flat/simplex/lexer/Token.java
package com.flat.simplex.lexer;

public class Token {

    private final String source;
    private final int start;
    private final int end;
    private final int length;
    private final Key key;
    private final String string;
    private Token next;
    private Token child;
    private Token lastChild;
    private int hash;

    public Token(String source, int start, int end, Key key) {
        this.source = source;
        this.start = start;
        this.end = end;
        this.length = end - start;
        this.key = key;
        if (key == Key.Invalid || key == Key.Word || key == Key.String || key == Key.Number) {
            this.string = source.substring(start, end);
        } else {
            this.string = key.name;
        }
    }

    public String getSource() {
        return source;
    }

    public int getStart() {
        return start;
    }

    public int getEnd() {
        return end;
    }

    public int getLength() {
        return length;
    }

    public Key getKey() {
        return key;
    }

    public String getString() {
        return string;
    }

    public Token getNext() {
        return next;
    }

    public void setNext(Token next) {
        this.next = next;
    }

    public Token getChild() {
        return child;
    }

    public void setChild(Token child) {
        this.child = child;
    }

    public void setLastChild(Token lastChild) {
        this.lastChild = lastChild;
    }

    public Token getLastChild() {
        return lastChild;
    }

    @Override
    public String toString() {
        return string;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Token other = (Token) o;
        return key == other.key && string.equals(other.string);
    }

    @Override
    public int hashCode() {
        if (hash == 0 && length > 0) {
            for (int i = start; i < end; i++) {
                hash = 31 * hash + source.charAt(i);
            }
        }
        return hash;
    }
}

File: SimplexApp/simplex/lexer/Token.js
var Key;

class Token {
    constructor(source, start, end, key) {
        this.source = source;
        this.start = start;
        this.end = end;
        this.length = end - start;
        this.next = null;
        this.child = null;
        this.lastChild = null;
        this.hashValue = 0;
        this.key = key;

        if (this.key === Key.Invalid || this.key === Key.Word || this.key === Key.String || this.key === Key.Number) {
            this.string = this.source.substring(this.start, this.end);
        } else {
            this.string = this.key.name;
        }
    }

    getSource() {
        return this.source;
    }

    getStart() {
        return this.start;
    }

    getEnd() {
        return this.end;
    }

    getLength() {
        return this.length;
    }

    getKey() {
        return this.key;
    }

    getString() {
        return this.string;
    }

    getNext() {
        return this.next;
    }

    setNext(next) {
        this.next = next;
    }

    getChild() {
        return this.child;
    }

    setChild(child) {
        this.child = child;
    }

    setLastChild(lastChild) {
        this.lastChild = lastChild;
    }

    getLastChild() {
        return this.lastChild;
    }

    toString() {
        return this.string;
    }

    equals(other) {
        if (this === other) {
            return true;
        }
        if (other === null || !(other instanceof Token)) {
            return false;
        }
        return this.key === other.key && this.string === other.string;
    }

    hashCode() {
        if (this.hashValue === 0 && this.length > 0) {
            for (let i = this.start; i < this.end; i++) {
                this.hashValue = ((this.hashValue << 5) - this.hashValue) + this.source.charCodeAt(i);
                this.hashValue = this.hashValue & this.hashValue;
            }
        }
        return this.hashValue;
    }
}

module.exports = Token;
Key = require("simplex/lexer/Key.js");